internal/pkg/limit/limiter.go

```diff
@@ -2,38 +2,88 @@
+// Package limit provides the ability to rate limit the api server.
 package limit
 
 import (
-	"errors"
+	"net/http"
 	"time"
 
 	"github.com/elastic/fleet-server/v7/internal/pkg/config"
+	"github.com/julienschmidt/httprouter"
 
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
 	"golang.org/x/sync/semaphore"
 	"golang.org/x/time/rate"
 )
 
+// Limiter enforces rate limits for each API endpoint.
 type Limiter struct {
-	rateLimit *rate.Limiter
-	maxLimit  *semaphore.Weighted
+	checkin  *limiter
+	artifact *limiter
+	enroll   *limiter
+	ack      *limiter
+	status   *limiter
+	log      zerolog.Logger
 }
 
-type ReleaseFunc func()
+// Create a new Limiter using the specified limits.
+func NewLimiter(addr string, cfg *config.ServerLimits) *Limiter {
+	return &Limiter{
+		checkin:  newLimiter(&cfg.CheckinLimit),
+		artifact: newLimiter(&cfg.ArtifactLimit),
+		enroll:   newLimiter(&cfg.EnrollLimit),
+		ack:      newLimiter(&cfg.AckLimit),
+		status:   newLimiter(&cfg.StatusLimit),
+		log:      log.With().Str("addr", addr).Logger(),
+	}
+}
 
-var (
-	ErrRateLimit = errors.New("rate limit")
-	ErrMaxLimit  = errors.New("max limit")
-)
+// WhapCheckin wraps the checkin handler with the rate limiter and tracks statistics for the endpoint.
+func (l *Limiter) WrapCheckin(h httprouter.Handle, i StatIncer) httprouter.Handle {
+	return l.checkin.wrap(l.log.With().Str("route", "checkin").Logger(), zerolog.WarnLevel, h, i)
+}
+
+// WhapArtifact wraps the artifact handler with the rate limiter and tracks statistics for the endpoint.
+func (l *Limiter) WrapArtifact(h httprouter.Handle, i StatIncer) httprouter.Handle {
+	return l.artifact.wrap(l.log.With().Str("route", "artifact").Logger(), zerolog.DebugLevel, h, i)
+}
+
+// WhapEnroll wraps the enroll handler with the rate limiter and tracks statistics for the endpoint.
+func (l *Limiter) WrapEnroll(h httprouter.Handle, i StatIncer) httprouter.Handle {
+	return l.enroll.wrap(l.log.With().Str("route", "enroll").Logger(), zerolog.DebugLevel, h, i)
+}
+
+// WhapAck wraps the ack handler with the rate limiter and tracks statistics for the endpoint.
+func (l *Limiter) WrapAck(h httprouter.Handle, i StatIncer) httprouter.Handle {
+	return l.ack.wrap(l.log.With().Str("route", "ack").Logger(), zerolog.DebugLevel, h, i)
+}
 
-func NewLimiter(cfg *config.Limit) *Limiter {
+// WhapStatus wraps the checkin handler with the rate limiter and tracks statistics for the endpoint.
+func (l *Limiter) WrapStatus(h httprouter.Handle, i StatIncer) httprouter.Handle {
+	return l.status.wrap(l.log.With().Str("route", "status").Logger(), zerolog.DebugLevel, h, i)
+}
+
+// StatIncer is the interface used to count statistics associated with an endpoint.
+type StatIncer interface {
+	IncError(error)
+	IncStart() func()
+}
+
+type releaseFunc func()
+
+type limiter struct {
+	rateLimit *rate.Limiter
+	maxLimit  *semaphore.Weighted
+}
 
+func newLimiter(cfg *config.Limit) *limiter {
```

Comment:
I don't love that the higher level app specific limiter is mixed in with the lower level implementation.  Makes it harder to understand.  Perhaps two separate files.